### ISSUE TITLE
Let loss_func(empty_inputs, reduction="mean") return 0

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#50224]
## 🚀 Feature
A loss function called with empty inputs (tensors with zero elements) and `reduction=mean` currently returns nan. 
It might be better to return 0.
```
$ F.cross_entropy(torch.rand(0, 10), torch.empty(0, dtype=torch.long), reduction="mean")
Out[6]: tensor(nan)
```
Same for many other losses.

## Motivation

If we consider the function using its mathematical definition, then the "mean" of losses is indeed undefined. ("sum" and "none" reduction do return reasonable results).
However if we consider the primary use of these functions as losses, returning 0 seems more reasonable.

## Alternatives

```
def cross_entropy(input, target, *, reduction="mean", **kwargs):
    if target.numel() == 0 and reduction == "mean":
        return input.sum() * 0.0  # connect the gradient
    return F.cross_entropy(input, target, **kwargs)
```

related: #12013

**Comment by gchanan** [comment]
I'd like to hear other's opinions, but I'm personally not in favor.

1) this is consistent with the definition of mean in the torch package and NumPy:
```
>>> torch.randn(0).mean()
tensor(nan)
>>> torch.randn(0, 1).mean(0)
tensor([nan])
```

you can of course redefine things in different contexts, but I find this sort of definition complexity gets out of control pretty quickly.

2) it's much more likely someone will find a NaN and be able to reason through what is happening when they wanted a 0 than the counterfactual case of returning a 0 when you wanted a NaN / didn't mean to pass an empty tensor.  The later is just much more likely to silently skate through.

**Comment by gchanan** [comment]
triage review to discuss the labeling and thoughts.

**Comment by albanD** [comment]
I do agree that mean should return `nan` for empty Tensors and it is consistent with numpy:
```python
import torch
import numpy as np

a = torch.rand(0, 10)
print(a.mean()) # Returns nan
print(np.mean(a.numpy())) # Returns nan
```

**Comment by ezyang** [comment]
Related-ish https://github.com/pytorch/pytorch/issues/50122

**Comment by jbschlosser** [comment]
I agree with @gchanan and @albanD that we should maintain the current behavior, for the same reasons:
1. Consistency with numpy
2. Minimizing the possibility of silently incorrect behavior

It's easy to manually work around the NaNs when 0 is preferred.

**Comment by albanD** [comment]
Closing this as there seems to be a consensus.
Feel free to re-open if there is a strong argument against it!

**Comment by ppwwyyxx** [comment]
Thanks for the feedbacks! Actually I think the behavior of `torch.mean` is only loosely relevant because no law says `F.cross_entropy` has to use `torch.mean` in the end. I think it is relevant, only in the sense that the name is similar, so an inconsistency will surprise users.

That reminds me of what tensorflow did - they just call it something else, a new `SUM_OVER_BATCH_SIZE` mode: https://www.tensorflow.org/api_docs/python/tf/compat/v1/losses/Reduction
Does it sound better to propose a new mode?

**Comment by ppwwyyxx** [comment]
FYI: the need is common enough so a wrapper of loss functions is created at https://github.com/facebookresearch/detectron2/commit/04fc85a0c44675559c2fbc9c7541cbb8b443819c to automatically return 0 when inputs are empty.

**Comment by jbschlosser** [comment]
Related: #70348
